File: kotlin/pig-latin/src/main/kotlin/PigLatin.kt
object PigLatin {
    fun translate(input: String) =
        if (vowelRegex.matches(input)) vowelRegex.replace(input, vowelReplacement)
        else consonantRegex.replace(input, consonantReplacement)

    private val vowelRegex = Regex("""(?<begin>^|\s+)(?<vowel>a|e|i|o|u|yt|xr)(?<rest>\w+)""")
    private val consonantRegex = Regex("""(?<begin>^|\s+)(?<consonant>ch|qu|thr|th|sch|yt|\wqu|\w)(?<rest>\w+)""")

    private const val vowelReplacement = "\${begin}\${vowel}\${rest}ay"
    private const val consonantReplacement = "\${begin}\${rest}\${consonant}ay"
}

File: javascript/pig-latin/pig-latin.js
const vowelRegex = RegExp(/(^|\s+)(a|e|i|o|u|yt|xr)(\w+)/g);
const consonantRegex = RegExp(/(^|\s+)(ch|qu|thr|th|sch|yt|\wqu|\w)(\w+)/g);

const vowelReplacement = "$1$2$3ay";
const consonantReplacement = "$1$3$2ay";

export default class PigLatin {
    translate(sentence) {
        if (vowelRegex.test(sentence))
            return sentence.replace(vowelRegex, vowelReplacement);
        else
            return sentence.replace(consonantRegex, consonantReplacement);
    }
}